test/e2e_node/cpu_manager_test.go

```diff
@@ -352,21 +362,123 @@ func runGuPodQuotaTest(ctx context.Context, f *framework.Framework, cpuCount int
 			gomega.Expect(quotas).To(gomega.Or(gomega.Equal("max"), gomega.Equal("-1")), "expected quota == max, got %q", quotas)
 		} else {
-			gomega.Expect(quotas).To(gomega.Not(gomega.Or(gomega.Equal("max"), gomega.Equal("-1"))), "expected quota to be enabled, got %q", quotas)
+			gomega.Expect(quotas).ToNot(gomega.Or(gomega.Equal("max"), gomega.Equal("-1")), "expected quota to be enabled, got %q", quotas)
 		}
 	}
 
+	// TODO Make sure the cpu manager reconcile loop executed already?
+
 	ginkgo.By(fmt.Sprintf("validating the cpu quota on Gu quota pod %s", pod.Name))
 
-	// TODO cpu manager reconcile period?
-	// podCgroupName, _ := m.GetPodContainerName(pod)
-	// containerConfig, _ := m.cgroupManager.GetCgroupConfig(podCgroupName, v1.ResourceCPU)
+	// Find the pinned container pid using the CRI Info interface
+	// This only works on linux
+	runtime, _, err := getCRIClient()
+	framework.ExpectNoError(err)
+	containers, err := runtime.ListContainers(context.Background(), &runtimeapi.ContainerFilter{
+		State: &runtimeapi.ContainerStateValue{
+			State: runtimeapi.ContainerState_CONTAINER_RUNNING,
+		},
+	})
+	framework.ExpectNoError(err)
+
+	// Identify the container
+	var guContainer *runtimeapi.Container
+	for _, c := range containers {
+		if c.Metadata.Name == "gu-container-quota" {
+			guContainer = c
+		}
+	}
+	gomega.Expect(guContainer).ToNot(gomega.Equal(nil), "could not find the pinned container via CRI")
+
+	// Retrieve the info block
+	containerStatus, err := runtime.ContainerStatus(context.Background(), guContainer.Id, true)
+	framework.ExpectNoError(err)
+	guPodInfo, ok := containerStatus.Info["info"]
+	gomega.Expect(ok).To(gomega.BeTrue(), "could not find info of the pinned container in %v", containerStatus.Info)
+
+	// The info block is retrieved in JSON format
+	// Search for a top level "pid" field
+	framework.Logf("Info: '%v'", guPodInfo)
+	info := InfoPid{}
+	err = json.Unmarshal([]byte(guPodInfo), &info)
+	framework.ExpectNoError(err)
+	gomega.Expect(info.Pid).ToNot(gomega.Equal(0), "could not find pid of the pinned container in %v", guPodInfo)
+	framework.Logf("Pinned process pid: %d", info.Pid)
+
+	// Use linux /proc APIs to detect the cgroup structure for the cpuset controller
+	cpusetRootPath, cpusetPath, cpuQuotaFileName, err := getLocalCpusetRootPath(info.Pid)
+	framework.ExpectNoError(err)
+
+	// Check quota in the container's and all parent cgroups to make sure there is no cgroup
+	// level that would introduce cpu throttling.
+	for {
+		cpuQuotaPath := path.Join(cpusetRootPath, cpusetPath, cpuQuotaFileName)
+		framework.Logf("Cpu quota path: %s", cpuQuotaPath)
+
+		out, err := exec.Command("sudo", "cat", cpuQuotaPath).Output()
+		if err == nil {
+			quotas := string(out)
+			gomega.Expect(quotas).To(gomega.Or(gomega.HavePrefix("max"), gomega.HavePrefix("-1")), "expected quota == max, got %q", quotas)
+		} else {
+			framework.Logf("could not read the cpu quota file: %s: %v", cpuQuotaPath, err)
```

Comment:
should we hard fail here? If not, why not?

Reply:
Missing cpu quota controller is a valid scenario afaik. Not every level has to have it enabled and so we can continue to the upper dir.